### PR TITLE
DO NOT MERGE - for comparison purposes

### PR DESCRIPTION
**Pull request by kyle-pena-kuzco**
## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: VERIFICATION_README.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
+# Verification
```
Comment (kyle-pena-kuzco) [inline VERIFICATION_README.md:1]: @samheutmaker - there's about 330 lines of other "stuff" like test-scripts and readmes, leaving about 440 lines of actual code changes. 

This includes:
- Feature: Generation of toploc "fingerprints" in both CUDA_GRAPH and regular model
- Feature: Verification of toploc fingerprints
- Feature: Ability to return input token IDs through API - important for exact repro, if, say, an input pre-processing step changes in sglang's impl
- Feature: Ability to return output token IDs

However, most of the code itself is just "pass this property along to the next class" glue type code.  There's a ton of that, and it closely tracks with the style / implementation of other features in the codebase, where possible.

Comment (kyle-pena-kuzco) [reply VERIFICATION_README.md:1]: I sincerely wish we could cut down on the size of the diff - and we can, if we're willing to offload some of it into our worker implementation.  However, that breaks our golden rule of workers being very dumb and not having their own logic.

 However, I'm still confident that we can keep this PR in sync because like I was saying 70%+ of it is just "pass along the property value to the next stage in the pipeline" type code.  Pretty easy to maintain.

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 2]
```diff
@@ -447,21 +465,29 @@ def run_once():
...
+        )
+        capture_hidden_mode_by_priority = {
+            v: k for k, v in capture_hidden_mode_priority.items()
+        }
+        highest_needed_capture_mode = capture_hidden_mode_by_priority[
+            max_priority_level
+        ]
+
+        if self.capture_hidden_mode != highest_needed_capture_mode:
+            self.capture_hidden_mode = highest_needed_capture_mode
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/model_executor/cuda_graph_runner.py:490]: This is really the highest risk change out of the whole fork.  The original code seemed to have an implicit "priority system" for picking capture mode in order to determine if the CUDA graph had to be re-captured prior to replay, and it was unclear how to extend this logic to include the toploc verification case, which requires either CaptureHiddenMode of FULL or LAST.  Weirdly, if there was no spec decoding and ForwardBatch.CaptureHiddenMode was LAST, it would switch it back to NONE.  I set ForwardBatch.CaptureHiddenModel to LAST for the sake of Toploc verification, so this code was "undoing" it.  That felt like a bug to me, so that specific case has changed.

File: python/pyproject.toml [hunk 3]
```diff
@@ -39,6 +39,7 @@ runtime_common = [
     "uvicorn",
     "uvloop",
     "xgrammar==0.1.16",
+    "toploc==0.1.4"
```
Comment (kyle-pena-kuzco) [inline python/pyproject.toml:42]: we use the open source implementation of fingerprint generation and verification provided by the paper authors.  why re-invent their wheel?

File: python/sglang/srt/layers/logits_processor.py [hunk 4]
```diff
@@ -54,6 +54,9 @@ class LogitsProcessorOutput:
     # The last hidden layers
     hidden_states: Optional[torch.Tensor] = None
 
+    # Captured hidden states for the purposes of generating and validating TopLOC "fingerprints", aka "fingerprints"
+    toploc_verification_hidden_states: Optional[torch.Tensor] = None
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/layers/logits_processor.py:58]: The `logits_processor` is the object that takes the last layer's output and processes it into token logits.  We need to store this here in order to process it into fingerprints.  EAGLE does something very simliar (see `hidden_states` just above)

File: python/sglang/srt/layers/logits_processor.py [hunk 5]
```diff
@@ -79,7 +82,7 @@ class LogitsProcessorOutput:
 class LogitsMetadata:
     forward_mode: ForwardMode
     capture_hidden_mode: CaptureHiddenMode = CaptureHiddenMode.NULL
-
+    toploc_verification: bool = False
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/layers/logits_processor.py:85]: Feature flag.  Needs to get passed around all over the place, unfortunately.

File: python/sglang/srt/layers/logits_processor.py [hunk 6]
```diff
@@ -347,11 +351,20 @@ def forward(
             else:
                 assert False, "Should never reach"
 
+        # If toploc is enabled, capture pruned hidden states
+        toploc_verification_hidden_states_to_store: Optional[torch.Tensor] = None
+        if logits_metadata.toploc_verification:
+            toploc_verification_hidden_states_to_store = (
+                pruned_states[sample_indices] if sample_indices else pruned_states
+            )
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/layers/logits_processor.py:359]: logits_processor contains  information at the batch level, so basically the first dimension has all the inferences in the current batch concatenated together.  each inference in a batch is called a sequence, "seq".

Thus if seq 1 has k tokens, and seq 2 has m tokens, then pruned_states has the hidden states for indices (k-1) and (k+m-1), which are the last tokens for seq 1 and seq 2.

In some cases there's only 1 sequence in a batch, but not in all cases.

Slicing out the hidden states for the last token in each sequence is what `pruned_states` is.  There's also this `sample_indices` but that has to do with more exotic usages of sglang that I don't currently think are relevant.

File: python/sglang/srt/managers/detokenizer_manager.py [hunk 7]
```diff
@@ -230,6 +230,10 @@ def handle_batch_token_id_out(self, recv_obj: BatchTokenIDOut):
             output_token_ids_logprobs_val=recv_obj.output_token_ids_logprobs_val,
             output_token_ids_logprobs_idx=recv_obj.output_token_ids_logprobs_idx,
             output_hidden_states=recv_obj.output_hidden_states,
+            toploc_verification_fingerprints=recv_obj.toploc_verification_fingerprints,
+            toploc_verification_fingerprint_validation_results=recv_obj.toploc_verification_fingerprint_validation_results,
+            origin_input_ids=recv_obj.origin_input_ids,
+            output_token_ids=recv_obj.output_token_ids,  # Copy output_token_ids from BatchTokenIDOut
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/detokenizer_manager.py:236]: passthru stuff here and elsewhere

File: python/sglang/srt/managers/io_struct.py [hunk 8]
```diff
@@ -475,6 +495,15 @@ class BatchStrOut:
...
 
+    # TopLOC Verification fingerprints
+    toploc_verification_fingerprints: List[List]
+    toploc_verification_fingerprint_validation_results: List[Optional[str]] = None
+
+    # Origin input ids (for return_input_ids=True)
+    origin_input_ids: Optional[List[List[int]]] = None
+    # Output token ids (for return_output_ids=True)
+    output_token_ids: Optional[List[List[int]]] = None
+
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/io_struct.py:506]: `GenerateReqInput` represents either a batch or a single request, depending on the context it's used in.  I'm following the pattern here as is used for all the other fields.

File: python/sglang/srt/managers/schedule_batch.py [hunk 9]
```diff
@@ -1313,6 +1331,25 @@ def get_model_worker_batch(self) -> ModelWorkerBatch:
...
+            else (
+                getattr(self.spec_info, "capture_hidden_mode", CaptureHiddenMode.NULL)
+                if self.spec_info
+                else CaptureHiddenMode.NULL
+            )
+        )
+
+        # Ensure CAPTURE_HIDDEN_MODE is *at least* LAST if toploc verification is enabled
+        if self.toploc_verification and capture_hidden_mode == CaptureHiddenMode.NULL:
+            capture_hidden_mode = CaptureHiddenMode.LAST
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/schedule_batch.py:1347]: CaptureHiddenMode has to be at least LAST in order for toploc verification to work, because otherwise we can't capture the last layer's activations in order to generate or validate a fingerprint.  CaptureHiddenMode signals to pytorch to retain these values so they can be cloned to the CPU.

File: python/sglang/srt/managers/scheduler_output_processor_mixin.py [hunk 10]
```diff
@@ -114,9 +122,31 @@ def process_batch_result_prefill(
...
+                        req.toploc_verification_fingerprints.append(
+                            create_toploc_fingerprint(toploc_verification_hidden_state)
+                        )
+                        if req.toploc_verification_fingerprint_to_validate is not None:
+                            req.toploc_verification_fingerprint_validation_result = (
+                                verify_toploc_fingerprint(
+                                    toploc_verification_hidden_state,
+                                    req.toploc_verification_fingerprint_to_validate,
+                                )
+                            )
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/scheduler_output_processor_mixin.py:149]: You'll see a very similar block of code relating to `hidden_states` here as well.  The difference is that I'm invoking the fingerprint and/or fingerprint verification methods depending on what is being requested.

File: python/sglang/srt/managers/scheduler_output_processor_mixin.py [hunk 11]
```diff
@@ -114,9 +122,31 @@ def process_batch_result_prefill(
                             .tolist()
                         )
 
-                    if req.grammar is not None:
-                        req.grammar.accept_token(next_token_id)
-                        req.grammar.finished = req.finished()
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/scheduler_output_processor_mixin.py:119]: I think this shouldn't have been dropped?  this may have been an oversight or a side effect of syncing with the main branch.

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 12]
```diff
@@ -919,6 +941,47 @@ def _handle_batch_output(
...
+                            ]
+                        )
+                        meta_info[
+                            "toploc_verification_fingerprint_validation_result"
+                        ] = validation_result
+                except Exception as e:
+                    logger.error(
+                        f"Error processing toploc verification fingerprint validation results: {e}"
+                    )
+
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/tokenizer_manager.py:984]: This looks like a lot, but it's all just what's required to pass things through in the API layer.

File: python/sglang/srt/server_args.py [hunk 13]
```diff
@@ -1063,6 +1066,18 @@ def add_cli_args(parser: argparse.ArgumentParser):
...
+            action="store_true",
+            help="Enable features relating to toploc verification",
+        )
+        parser.add_argument(
+            "--toploc-verification-topk",
+            type=int,
+            default=128,
+            help="Top-k for TopLoc verification",
+        )
+
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/server_args.py:1080]: This is where we specify the flags.

File: python/sglang/srt/managers/scheduler_output_processor_mixin.py [hunk 14]
```diff
@@ -250,6 +280,30 @@ def process_batch_result_decode(
...
+                            verify_toploc_fingerprint(
+                                toploc_verification_hidden_state[-1, ...],  # last token
+                                req.toploc_verification_fingerprint_to_validate,
+                            )
+                        )
+                else:
+                    # No need to generate a fingerprint until the last decode step of the sequence
+                    req.toploc_verification_hidden_states.append(None)
+                    req.toploc_verification_fingerprints.append(None)
+
```
Comment (kyle-pena-kuzco) [inline python/sglang/srt/managers/scheduler_output_processor_mixin.py:306]: `process_batch_result_decode` is where "post processing" on a "decode" happens.  A "decode" is the production of a single new token.  We use the `req.finished` flag to only do toploc stuff on the last token.
